apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -133,35 +211,158 @@ type ReadinessCheck struct {
+	// the list of Patches on a resource.
+	// +optional
+	ID *string `json:"id,omitempty"`
+
+	// NOTE(bagricola): We specify PatchSetFromCompositeFieldPath here because
+	// we need to specify a type which is not simply Patch. If we specify
+	// Patch, we're basically making a recursive structure which kubebuilder
+	// does not like. We don't want to nest PatchSets inside PatchSets anyway,
+	// so we specify the only other available type (right now).
+	// Once more Patch Types are added, this should be a composite type which
+	// encompasses all of the other 'non Patch Set' Patch Types.
+
+	// Patches defines the list of Patch objects which are part of this PatchSet
+	Patches []PatchSetCompatiblePatch `json:"patches,omitempty"`
+}
+
+// Apply applies each patch in a set.
+func (c *PatchSet) Apply(from, to runtime.Object) error {
+	// PatchSets must have an ID
+	if c == nil || c.ID == nil {
+		return errors.New(errInvalidPatch)
+	}
+
+	for i, p := range c.Patches {
+		if err := p.Apply(from, to); err != nil {
+			return errors.Wrapf(err, errInvalidPatchSetPatch, i, *c.ID)
+		}
+	}
+	return nil
+}
+
+// PatchFromCompositeFieldPath is a standard patch - taking an input value
+// from the composite resource, and applying it to the composed resource,
+// optionally transforming it along the way.
+type PatchFromCompositeFieldPath struct {
 	// FromFieldPath is the path of the field on the upstream resource whose value
 	// to be used as input.
-	FromFieldPath string `json:"fromFieldPath"`
+	// +optional
+	FromFieldPath *string `json:"fromFieldPath,omitempty"`
 
 	// ToFieldPath is the path of the field on the base resource whose value will
 	// be changed with the result of transforms. Leave empty if you'd like to
 	// propagate to the same path on the target resource.
 	// +optional
-	ToFieldPath string `json:"toFieldPath,omitempty"`
+	ToFieldPath *string `json:"toFieldPath,omitempty"`
 
 	// Transforms are the list of functions that are used as a FIFO pipe for the
 	// input to be transformed.
 	// +optional
 	Transforms []Transform `json:"transforms,omitempty"`
 }
 
-// Apply runs transformers and patches the target resource.
-func (c *Patch) Apply(from, to runtime.Object) error {
+// Apply patches the composed resource, using fields from the composite resource.
+// Values may be transformed if any are defined on the patch.
+func (c *PatchFromCompositeFieldPath) Apply(from, to runtime.Object) error { // nolint:gocyclo
+	// NOTE(benagricola): The cyclomatic complexity here is from error checking
+	// at each stage of the patching process, in addition to Apply methods now
+	// being responsible for checking the validity of their input fields
+	// (necessary because with multiple patch types, the input fields
+	// must be +optional).
+	if c == nil || c.FromFieldPath == nil {
+		return errors.New(errInvalidPatch)
+	}
+
+	// Default to patching the same field on the composed resource
+	if c.ToFieldPath == nil {
+		c.ToFieldPath = c.FromFieldPath
+	}
```

Comment:
Is this functionality new? I like it, but we should make sure to document it.

Reply:
@negz `ToFieldPath` is actually already advertised as working this way (see description of ToFieldPath in Patch), but not implemented. I was literally just about to make a separate PR with just that :smiley:

Reply:
Yeah - I actually implemented this because I already believed from the doco that it worked that way, and I'd only broken it by making all of the fields optional + string pointers 😆

Reply:
Well, that's convenient! Thanks for the update. :)